[WIP] Adding Sharelatex integration

danielneis opened the pull request:
Hello,

we're developing an integration with ShareLaTeX (www.sharelatex.com) as a storage provider.

We already made some progress to get the plugin saving and deleting the credentials and also listing projects from sharelatex, saving a project from sharelatex to a project in OSF and listing the file tree of a ShareLatex project on OSF.

Now we are running the tests but it seems that we are not correctly "mocking" the ShareLatex service and would appreciate some advices on this.

Kind regards,
Daniel

danielneis commented:
Hello,

we're trying to make the "view" action of our addon to work, and altough the download action is working, it seems that the behaviour for view action is different.
Here is the link for the gist with the "error" when trying to "view" a file from our addon:https://gist.github.com/danielneis/5664b0a831bf94920852
The problems seems to be that OSF tries to authenticate, then get the file (what results in a 302 HTTP redirect) and then tries to authenticate again.

Also, it seems that it is calling waterbutler for both actions, but with very different URLs as seem on the following gist (the two first lines are for view and the last one for download)
https://gist.github.com/danielneis/d8b76507b19e076f54e4

lbanner commented:
Following up @danielneis -- appreciate the PR ... @mfraezz ... This is add-on work, can you give a review and feedback? See notes above. Thanks!

lbanner commented:
@danielneis I'm wondering if if might be possible for you to write up a design doc for the Sharelatex add-on and share it via link here? What would be helpful is an articulation of the functional and user expectations, the use-case, as well as a wireframe or screenshots from your local environment of how a user is meant to interact with the add-on via the OSF -- This is valuable for both code-review and QA.

What do you think? Thanks!

danielneis commented:
@lbanner 

We've added some text to the "help" section.
It is mostly a read-only plugin to allow people to add it as a Storage Addon and browse the files inside OSF.
I've made some screenshots, but not added it to the docs already. Will do it soon.

Kind regards,
Daniel

felliott commented:
Hey @danielneis,

Where on ShareLaTeX do I get my auth token?  I've created an account, but I can't find anything on the site.

Cheers,
@felliott

danielneis commented:
Hello,

as our work on the API is still in progress awaiting for review by the ShareLatex team (https://github.com/sharelatex/web-sharelatex/pull/212), you can access your token going to https://www.sharelatex.com/user/auth_token
When the new API got integrated, it will be visible on profie page.
Also, the integration with OSF will not work with the site available at sharelatex.com, you will have to install and run your own instance to test it.

Kind regards,
Daniel

danielneis commented:
I'll use this comment to note which tests are failing:

1) Must take a closer look.

sharelatexNodeConfigViewModel
   ViewModel
      ✗ imports default settings if not given during instantiation
	AssertionError: expected undefined to equal true (/home/travis/build/CenterForOpenScience/osf.io/website/static/js/tests/tests.webpack.js:28794:0 <- webpack:///~/chai/chai.js:875:0)

2) We must check if this is the actual behaviour. Maybe we have changed the plugin and must fix the test.

 #toggleBucket
      ✗ shows the project selector when disabled and if projects aren't loaded fetches the list of projects
	TypeError: Fake XHR onreadystatechange handler threw exception: 'undefined' is not a function (evaluating 'vm.showSelect(false)') (/home/travis/build/CenterForOpenScience/osf.io/website/static/js/tests/tests.webpack.js:127021:0 <- webpack:///website/addons/sharelatex/static/tests/sharelatexNodeConfig.test.js:311:20)

3) It is expected as the ShareLatex API does not expose a method to create projects.
Should we remove this test?

#createBucket
      ✗ sends a POST to create project and on success updates the project list
	TypeError: Fake XHR onreadystatechange handler threw exception: 'undefined' is not a function (evaluating 'vm.createBucket(name)') (/home/travis/build/CenterForOpenScience/osf.io/website/static/js/tests/tests.webpack.js:127208:0 <- webpack:///website/addons/sharelatex/static/tests/sharelatexNodeConfig.test.js:515:20)

felliott commented:
Minor: could you add a `README.md` to the `website/addons/sharelatex/` directory.  It should explain how to get an auth token and set up sharelatex for local development.  See the README.md in the googledrive addon for a good example.

felliott commented:
This isn't working for me ATM.  `website.addons.sharelatex.utils.get_project_list` is trying to get the contents of `http://localhost:3000/api/v1/project?auth_token=myauthtoken` and is getting the HTML 'Page Not Found' error.

felliott commented on website/addons/sharelatex/utils.py:
```diff
@@ -0,0 +1,88 @@
+import urllib2
+import httplib
+
+from boto import exception
```
`boto` is a library for interacting with Amazon AWS services.  Why are you using that here?

danielneis commented:
Hello, @felliott 

I've removed the boto imports (we've copied them from the amazon plugin and forgot to remove) and also added a readme poiting to the sharelatex documentation on install and also linking to the pull request that you will need to merge until the sharelatex people merge it to mainline.

kind regards,
Daniel

felliott commented:
Hey @danielneis,

Thank you for updating the README.md.  I've run into one more issue with connecting ShareLaTeX to the OSF.  On the user's Configure Add-on page I originally put in `http://localhost:3000/` as the ShareLaTeX URL:

![screen shot 2016-03-15 at 1 43 31 pm](https://cloud.githubusercontent.com/assets/86835/13787947/677f2ae0-eab4-11e5-8c39-ef302c9b1046.png)

The trailing slash at the end of the URL breaks the `website.addon.sharelatex.utils.get_project_list` function because another slash is added between the domain and the path.  My url ends up being `http://localhost:3000//api/v1/project?auth_token=...` with the double-slash causing it to 404.  I can work around this for now by updating the sharelatex url in my users settings to not have the slash, but the url should be normalized when the user adds it.

My local dev server and WB are now communicating, so I should be able to start CRing the WB pull request.

danielneis commented:
hello, @felliott 

we've made some changes that should have fixed the problem with the URL.
We still do not have a test for that but are planning to do so on the next round.

Kind regards,
Daniel

brianjgeiger commented:
It's Autumn Cleaning time, so I'm closing this PR. Feel free to re-open when development becomes active again.

vals commented:
Did anything happen with this?

jamescdavis commented:
@vals it appears the [PR to add a public API to ShareLaTeX](https://github.com/sharelatex/web-sharelatex/pull/212) was held up due to issues with private tokens and later closed without merging, probably due to inactivity. You would need to start there as a public API would be required for OSF integration. I'd try contacting @danielneis to find out the status of things and see if you can help.
